cou/steps/plan.py

```diff
@@ -16,25 +16,94 @@
 """Upgrade planning utilities."""
 
 import logging
+from collections import defaultdict
+from typing import List
 
 from cou.steps import UpgradeStep
 from cou.steps.analyze import Analysis
+from cou.steps.application.app import StandardApplication
 from cou.steps.backup import backup
+from cou.utils.openstack import (
+    SPECIAL_CHARMS,
+    UPGRADE_ORDER,
+    CompareOpenStack,
+    determine_next_openstack_release,
+)
 
 logger = logging.getLogger(__name__)
 
 
-async def generate_plan(args: Analysis) -> UpgradeStep:
+async def generate_plan(analysis_result: Analysis) -> UpgradeStep:
```

Comment:
We can do like that however plan makes generics things easier. Such as:
- Asking for confirmation on each step.
- Marking steps as serial or parallel without changing the executors
- For example there is a request to see how long each step took. 
- Dumping the overall upgrade plan
- ...

Reply:
Yes, having the applications only return libjuju commands back to the execute step does make tweaking the execution easier, but I don't think it'd be impossible to do it within the application classes. For example:
* confirmation can be given from the execute step before running any (potentially asynchronous batch of) `Application.upgrade()` calls
* I'd be in favor of a single predetermined way to upgrade a given group of apps; too much complexity otherwise
* timing execution could be done within the Application class
* printing out the upgrade plan could be done by the Application classes logging their behavior + a dry-run parameter in the `upgrade()` method

Reply:
But those are not the responsibility of Application itself. It just need to know how to upgrade application or unit.